src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatisticsCollector.cs

```diff
@@ -0,0 +1,169 @@
+            switch (peerEvent)
+            {
+                case Event.PeerConnected @event:
+                    statistics.Inbound = @event.Inbound;
+                    statistics.LogEvent($"Peer Connected");
+                    break;
+                case Event.PeerConnectionAttempt @event:
+                    statistics.Inbound = @event.Inbound;
+                    statistics.LogEvent($"Attempting Connection");
+                    break;
+                case Event.PeerConnectionAttemptFailed @event:
+                    statistics.Inbound = @event.Inbound;
+                    statistics.LogEvent($"Connection attemp FAILED. Reason: {@event.Reason}.");
+                    break;
+                case Event.PeerDisconnected @event:
+                    statistics.Inbound = @event.Inbound;
+                    statistics.LogEvent($"Disconnected. Reason: {@event.Reason}. Exception: {@event.Exception?.ToString()}");
+                    break;
+                case Event.PeerMessageReceived @event:
+                    statistics.ReceivedMessages++;
+                    statistics.BytesReceived += @event.Message.MessageSize;
+                    statistics.LogEvent($"Message Received: {@event.Message.Payload.Command}");
+                    break;
+                case Event.PeerMessageSent @event:
+                    statistics.SentMessages++;
+                    statistics.BytesSent += @event.Size;
+                    statistics.LogEvent($"Message Sent: {@event.Message.Payload.Command}");
+                    break;
+                case Event.PeerMessageSendFailure @event:
+                    statistics.LogEvent($"Message Send Failure: {@event.Message?.Payload.Command}. Exception: {@event.Exception?.ToString()}");
+                    break;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private PeerStatistics GetPeerStatistics(IPEndPoint peerEndPoint)
+        {
+            if (!this.peersStatistics.TryGetValue(peerEndPoint, out PeerStatistics statistics))
+            {
+                // ensures no other threads have created already an entry between existance check and lock acquisition.
+                if (!this.peersStatistics.TryGetValue(peerEndPoint, out statistics))
+                {
+                    statistics = new PeerStatistics(this.diagnosticSettings.MaxPeerLoggedEvents, peerEndPoint);
+                    this.peersStatistics.Add(peerEndPoint, statistics);
+                }
+            }
+
+            return statistics;
+        }
+
+
+        public void StartCollecting()
+        {
+            lock (this.lockStartStopCollecting)
+            {
+                this.peersStatistics.Clear();
+
+                this.peersEventsQueue = this.asyncProvider.CreateAndRunAsyncDelegateDequeuer<Event.PeerEventBase>(nameof(this.peersEventsQueue), UpdatePeerStatistics);
+
+                this.eventSubscriptions.Add(this.signals.Subscribe<Event.PeerConnected>(this.EnqueuePeerEvent));
```

Comment:
Is there maybe a way that we can just subscribe to all the `PeerEventBase` classes? Or do you want to be explicit.

Reply:
We want to be explicit using event bus, will prevent to subscribe to unneeded events and at the same time expresses what we handle